### ISSUE TITLE
Debugging the OpenFAST

### ISSUE THREAD

**Issue by selvavm** [OpenFAST/openfast#211]
Hi,

I have recently started working in OpenFAST and I am compiling it in linux with gcc@7.3.0

I managed to build it but I am facing Invalid Memory reference. So, I want to debug the code. Is there any document which mentions on how to setup the dev environment for debugging and maybe add features?

Although my final version will be in linux, I can also use Windows to do development so I tried VS 2015 (vs-build) but it says unrecognised project (`The application which this project type is based on was not found.`). Moreover, I can't get Intel FORTRAN compiler for my windows machine:( 

PS: I am a beginner to FORTRAN and C

**Comment by rafmudaf** [comment]
Hi @selvavm 

Thanks for getting in touch about developing and debugging OpenFAST. The debugging process will be a little different for each system, but I can explain what I do on linux.

First, compile OpenFAST in debug mode by setting `CMAKE_BUILD_TYPE` to `Debug`. You can do this on the command line with
```
cmake .. -D CMAKE_BUILD_TYPE=Debug
```
or by using `ccmake` to open the command line cmake gui to change it. While thats compiling, you can set up the rest of your debugging tools.

I primarily use the GNU debugger, `gdb`, and couple that with [Visual Studio Code](https://code.visualstudio.com) and its [Native Debug](https://marketplace.visualstudio.com/items?itemName=webfreak.debug) extension.

For Native Debug, you have to sort of hack the extension to allow breakpoints in fortran files by adding this line to your `settings.json`:
```
    "debug.allowBreakpointsEverywhere": true,
```
You can set up a launch configuration in VS Code so that you can debug a particular OpenFAST case through the IDE. To do this, open the [launch configuration](https://code.visualstudio.com/docs/editor/debugging#_launch-configurations) and add a block similar to this:
```
        {
            "name": "AOC_WSt",
            "type": "gdb",
            "request": "launch",
            "printCalls": false,
            "showDevDebugOutput": false,
            "valuesFormatting": "prettyPrinters",
            "gdbpath": "gdb",
            "target": "${workspaceRoot}/build/glue-codes/openfast/openfast",
            "cwd": "${workspaceRoot}/build/reg_tests/glue-codes/openfast/AOC_WSt/",
            "arguments": "${workspaceRoot}/build/reg_tests/glue-codes/openfast/AOC_WSt/AOC_WSt.fst",
        },
```

It may by helpful to write a small fortran program or clone this [square root program](https://github.com/rafmudaf/sqrt) of mine to verify that you can actually set breakpoints and inspect variables with your tools. You could do something you know is an error like access an array element that is not allocated and verify that VS Code shows you where the bug occurs. Using a small program means you can keep things simple while setting up your tooling. For example, recompiling the sqrt program takes less than a second while compiling OpenFAST takes ~15-20 minutes.

**Comment by rafmudaf** [comment]
I hope this gets you started and let me know if you have any further questions or comments. Other OpenFAST developers, @bjonkman @ghaymanNREL @gantech, do you have any useful tips for debugging?

The documentation does not currently have debugging tips so I'll leave this issue open with the action to add simple debugging tips.

**Comment by rafmudaf** [comment]
closed by #262
